### PR TITLE
Hierarchy dot string

### PR DESCRIPTION
**Pull request by ss2165**
for visualising hierarchy alongside portgraph with new function `hier_graph_dot_string_with`

image

### CODE REVIEW

File: src/dot.rs [hunk 1]
```diff
@@ -32,15 +82,29 @@ where
...
+    dot.push_str("digraph {\n");
+
+    let (dot_node, dot_edge) = node_and_edge_strings(graph, nodes, ports);
+    dot.push_str(&dot_node);
+    dot.push_str(&dot_edge);
+    dot.push_str("}\n");
+    dot
+}
+
+fn node_and_edge_strings(
```
Comment (aborgna-q) [inline src/dot.rs:100]: Add a docstring?

File: src/dot.rs [hunk 2]
```diff
@@ -2,11 +2,61 @@
...
+use crate::{Direction, Hierarchy, NodeIndex, PortGraph, PortIndex, Weights};
 
 /// Style of an edge in a dot graph. Defaults to "None".
 pub type DotEdgeStyle = Option<String>;
 
+/// Encode `PortGraph` and `Hierarchy` in dot format.
+///
+/// Calls the `nodes` and `ports` functions to get the labels for the nodes and ports.
+/// The `ports` function also returns the style of the edge connected to a port.
+///
```
Comment (aborgna-q) [inline src/dot.rs:14]:
```suggestion
/// Calls the `ports` function to get the labels for the ports and the style of
/// the edges connected to them.
```

File: src/dot.rs [hunk 3]
```diff
@@ -2,11 +2,61 @@
...
+    nodes: impl FnMut(NodeIndex) -> String,
+    ports: impl FnMut(PortIndex) -> (String, DotEdgeStyle),
+) -> String {
+    let mut dot = String::new();
+
+    dot.push_str("digraph {\n");
+
+    let (mut dot_node, mut dot_edge) = node_and_edge_strings(graph, nodes, ports);
+
+    // dot_node.push_str("digraph {\n");
```
Comment (aborgna-q) [inline src/dot.rs:29]:
```suggestion
```

File: src/dot.rs [hunk 4]
```diff
@@ -2,11 +2,61 @@
...
+
+    for n in graph.nodes_iter() {
+        let node_str = format!(
+            "{} [shape=plain label=\"{}\"]\n",
+            hier_node_id(n),
+            n.index()
+        );
+        dot_node.push_str(&node_str);
+
+        // Connect the parent to any existing first and last children
```
Comment (aborgna-q) [inline src/dot.rs:40]:
```suggestion
        // Connect the parent to any existing children
```
?

**Review by aborgna-q — COMMENTED** [review]
👍 

Returning `dot_node` from `node_and_edge_strings` adds an unnecessary strring copy when we prepend the "digraph {" line.

Perhaps the function could take &mut String, use it as dot_node, and append dot_edge at the end of the function? And then have the hierarchy graph appended to that?

File: src/dot.rs (hunk 2 above)
Comment (ss2165) [reply src/dot.rs:14]: I just copied the existing docstring for `dot_with...`

Comment (aborgna-q) [reply src/dot.rs:14]: Sorry, for some reason I though there was a duplicated line mentioning the nodes.

**Review by aborgna-q — APPROVED** [review]

**Pull request merged**